File: 0082_Remove_Duplicates_from_Sorted_List_II/solution.java
package cn.huang.leetcode;

import cn.huang.leetcode.common.ListNode;

/*
82. Remove Duplicates from Sorted List II
Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.

Example 1:

Input: 1->2->3->3->4->4->5
Output: 1->2->5

Example 2:

Input: 1->1->1->2->3
Output: 2->3


 */
public class LC_0082_RemoveDuplicatesfromSortedList_II {
    public ListNode deleteDuplicates(ListNode head) {
        if(head == null || head.next == null)
            return head;

        ListNode dummyNode = new ListNode(-1);
        ListNode prev = dummyNode;
        dummyNode.next = head;

        while(head != null && head.next != null)
        {
            if (head.val == head.next.val)
            {
                int value = head.val;
                while (head!=null && head.val == value)
                    head = head.next;
                prev.next = head;
            }
            else
            {
                head = head.next;
                prev = prev.next;
            }

        }
        return dummyNode.next;
    }

}

File: 0082_Remove_Duplicates_from_Sorted_List_II/solution.cpp
/*
82. Remove Duplicates from Sorted List II

Level: Medium

https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii
*/
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
  ListNode *deleteDuplicates(ListNode *head)
  {
    if (head == nullptr)
      return head;
    ListNode dummy(-1);
    dummy.next = head;

    ListNode *unique = &dummy;
    while (head != nullptr && head->next != nullptr)
    {
      // find deuplicate value, update value for delete
      if (head->val == head->next->val)
      {
        int value = head->val;
        while (head != nullptr && head->val == value)
          head = head->next;
        unique->next = head;
      }
      else
      {
        head = head->next;
        unique = unique->next;
      }
    }

    return dummy.next;
  }
};